Container and memory allocation guidelines (#12387)
microsoft/onnxruntime: docs/Coding_Conventions_and_Standards.md

  * Allowed
  * Use a non-const reference for arguments that are modifiable but cannot be `nullptr` so the API clearly advertises the intent
  * Const correctness and usage of smart pointers (`shared_ptr` and `unique_ptr`) is expected. A non-const reference equates to "this is a non-null object that you can change but are not being given ownership of".

> * Prefer passing `gsl::span<const T>` by value (or `std::span` when supported) as input arguments when passing const references to containers with contiguous storage (like `std::vector`). This allows to make the function container independent, represent arbitrary memory spans or pass sub-spans as an argument. The below examples allow the client code to use either `std::vector`, `InlinedVector`, an instance of a `gsl::span` would be created automatically

Reviewer: nit: wording suggestions
```suggestion
* Prefer passing `gsl::span<const T>` by value (or `std::span` when supported) as input arguments when passing const references to containers with contiguous storage (like `std::vector`). This allows the function to be container independent, and the argument to represent arbitrary memory spans or sub-spans. The below examples allow the client code to use either `std::vector` or `InlinedVector`. An instance of a `gsl::span` would be created automatically.
```

> ```

Author: Yes.

/// Instead of
void foo(const std::vector<int64_t>&);

/// Use to pass any contiguous const container containing int64_t

> // Now you can seamless pass either `std::vector`, `InlinedVector`, `std::array` or `gsl::span` as an argument.

Reviewer: ```suggestion
// Now you can seamlessly pass either `std::vector`, `InlinedVector`, `std::array` or `gsl::span` as an argument.
```

void foo(gsl::span<const int64_t>);

// Example with pointer to const data. Instead of
void foo(const std::vector<const Node*>&);

// Use
void foo(gsl::span<const Node* const>);
```

> * Prefer returning `gsl::span<const T>` by value instead of a const reference to a contiguous member container or memory span. This allows the client code to be container independent.

Reviewer: hm, I'm still not getting it - isn't a `span<const T>` also referring to a chunk of memory?

```cpp
// Original code
void foo(const std::vector<std::string>&);

foo({"abc", "dbf"}); // Works

// After refactoring to gsl::span it would no longer compile. Use AsSpan().
void foo(gsl::span<const std::string>);

foo(AsSpan<std::string>{"abc", "dbf"}); // Works
```
* Prefer passing `std::string_view` by value instead of `const std::string&`. Make sure that the lifespan of a `std::string` instance ecplises the lifespan of the corresponding `std::string_view` instance.

* `using namespace` permitted with limited scope
  * Not allowing `using namespace` at all is overly restrictive. Follow the C++ Core Guidelines:
    * [SF.6: Use using namespace directives for transition, for foundation libraries (such as std), or within a local scope (only)](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rs-using)
    * [SF.7: Don't write using namespace at global scope in a header file](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rs-using-directive)

### Containers to use

> Onnxruntime aims to reduce latency and latency variance by minimizing the amount of dynamic memory allocations and avoid using locks.

Reviewer: is "avoid using locks" relevant to this section or a more general guideline?

* Use `reserve()` on hash containers and vectors. For example,
```cpp
#include "core/common/inlined_containers.h"

void foo(gsl::span<const std::string> names) {
  // For local processing, names are still valid
  // use std::string_view to avoid duplicate memory allocations.
  // same code would work with std::unordered_set if built without Abseil
  InlinedHashSet<std::string_view> unique_names;
  unique_names.reserve(names.size());
  unique_names.insert(names.cbegin(), names.cend());
}
```

### Other

* Qualify usages of `auto` with `const`, `*`, `&` and `&&` where applicable to more clearly express the intent
* When adding a new class, disable copy/assignment/move until you have a proven need for these capabilities. If a need arises, enable copy/assignment/move selectively, and when doing so validate that the implementation of the class supports what is being enabled.
  * Use `ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE` initially
  * See the other `ORT_DISALLOW_*` macros in <https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/common/common.h>

> * `std::unique_ptr` is often used for delayed or optional construction of objects or members of classes. Use `std::optional` as appropriate to reduce the number of allocations.

Reviewer: just to convey more strongly that `optional` is the suggested alternative
```suggestion
* Sometimes, `std::unique_ptr` might be considered for delayed or optional construction of objects or members of classes. Instead, use `std::optional` as appropriate to reduce the number of allocations.
```

  * Const correctness and usage of smart pointers (`shared_ptr` and `unique_ptr`) is expected. A non-const reference equates to "this is a non-null object that you can change but are not being given ownership of".
* Prefer passing `gsl::span<const T>` by value (or `std::span` when supported) as input arguments when passing const references to containers with contiguous storage (like `std::vector`). This allows the function to be container independent, and the argument to represent arbitrary memory spans or sub-spans. The below examples allow the client code to use either `std::vector` or `InlinedVector`. An instance of a `gsl::span` would be created automatically.
```cpp
/// Instead of
void foo(const std::vector<int64_t>&);

/// Use to pass any contiguous const container containing int64_t
// Now you can seamlessly pass either `std::vector`, `InlinedVector`, `std::array` or `gsl::span` as an argument.
void foo(gsl::span<const int64_t>);

// Example with pointer to const data. Instead of
void foo(const std::vector<const Node*>&);

// Use
void foo(gsl::span<const Node* const>);
```

> * Prefer returning `gsl::span<const T>` by value instead of a const reference to a contiguous member container. Prefer returning `gsl::span` instead of a pointer referring to a chunk of memory. The size is also included in the span.

Reviewer: ```suggestion
* Prefer returning `gsl::span<const T>` by value instead of a const reference to a contiguous member container. Prefer returning `gsl::span` instead of a pointer referring to a chunk of memory as the size is also included in the span.
```
as suggested by @skottmckay
